core/src/connection_reuse.rs

```diff
@@ -281,145 +487,312 @@ where
+    C: ConnectionUpgrade<T::Output, T::MultiaddrFuture>,
+    C::Output: StreamMuxer + Clone,
 {
-    type Item = future::FutureResult<(M::Substream, future::FutureResult<Multiaddr, IoError>), IoError>;
+    type Item = future::FutureResult<(ConnectionReuseSubstream<T, C>, future::FutureResult<Multiaddr, IoError>), IoError>;
     type Error = IoError;
 
+    #[inline]
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        let mut lock = self.shared.lock();
-
-        // Try to get any new muxer from `add_to_next_rx`.
-        // We push the new muxers to a channel instead of adding them to `next_incoming`, so that
-        // tasks are notified when something is pushed.
-        loop {
-            match lock.add_to_next_rx.poll() {
-                Ok(Async::Ready(Some(elem))) => {
-                    lock.next_incoming.push(elem);
-                }
-                Ok(Async::NotReady) => break,
-                Ok(Async::Ready(None)) | Err(_) => unreachable!(
-                    "the sender and receiver are both in the same struct, therefore \
-                     the link can never break"
-                ),
-            }
+        let mut shared = self.shared.lock();
+        match poll_incoming(&self.shared, &mut shared, None) {
+            Ok(Async::Ready(Some(substream))) => {
+                Ok(Async::Ready(substream))
+            },
+            Ok(Async::Ready(None)) | Ok(Async::NotReady) => {
+                // TODO: will add an element to the list every time
+                shared.notify_on_new_connec.insert(TASK_ID.with(|&v| v), task::current());
+                Ok(Async::NotReady)
+            },
+            Err(err) => Err(err)
         }
+    }
+}
 
-        // Check whether any incoming substream is ready.
-        for n in (0..lock.next_incoming.len()).rev() {
-            let (muxer, mut future, addr) = lock.next_incoming.swap_remove(n);
-            match future.poll() {
-                Ok(Async::Ready(None)) => {
-                    debug!("no inbound substream for {}", addr);
-                    lock.active_connections.remove(&addr);
-                }
-                Ok(Async::Ready(Some(value))) => {
-                    // A substream is ready ; push back the muxer for the next time this function
-                    // is called, then return.
-                    debug!("New incoming substream");
-                    let next = muxer.clone().inbound();
-                    lock.next_incoming.push((muxer, next, addr.clone()));
-                    return Ok(Async::Ready(future::ok((value, future::ok(addr)))));
+/// Polls the incoming substreams on all the incoming connections that match the `listener`.
+///
+/// Returns `Ready(None)` if no connection is matching the `listener`. Returns `NotReady` if
+/// one or more connections are matching the `listener` but they are not ready.
+fn poll_incoming<T, C>(shared_arc: &Arc<Mutex<Shared<T, C>>>, shared: &mut Shared<T, C>, listener: Option<u64>)
```

Comment:
having both `shared_arc: &Arc<Mutex<Shared<T, C>>>` and  `shared: &mut Shared<T, C>` as parameters feels very odd, and makes me wonder if this isn't supposed to be somewhere else closed to `Shared` itself.